src/axom/core/Array.hpp

```diff
@@ -728,48 +837,65 @@ class Array
   }
   /// @}
 
-  T* m_data;
-  IndexType m_num_elements;
-  IndexType m_capacity;
-  double m_resize_ratio;
-  bool m_is_external;
+  T* m_data = nullptr;
+  /// \brief The full number of elements in the array
+  ///  i.e., 3 for a 1D Array of size 3, 9 for a 3x3 2D array, etc
+  IndexType m_num_elements = 0;
+  IndexType m_capacity = 0;
+  double m_resize_ratio = DEFAULT_RESIZE_RATIO;
+  bool m_is_external = false;
   int m_allocator_id;
+  // FIXME: What's the best way to make these available to users
+  // Is it sufficient to just give them the raw arrays?
+  /// \brief The sizes (extents?) in each dimension
+  IndexType m_dims[dim];
+  /// \brief The strides in each dimension
+  IndexType m_strides[dim];
 };
 
 //------------------------------------------------------------------------------
 //                            Array IMPLEMENTATION
 //------------------------------------------------------------------------------
 
 //------------------------------------------------------------------------------
-template <typename T>
-Array<T>::Array()
-  : m_data(nullptr)
-  , m_num_elements(0)
-  , m_capacity(0)
-  , m_resize_ratio(DEFAULT_RESIZE_RATIO)
-  , m_is_external(false)
-  , m_allocator_id(axom::getDefaultAllocatorID())
+template <typename T, IndexType dim>
+Array<T, dim>::Array() : m_allocator_id(axom::getDefaultAllocatorID())
 { }
 
+template <typename T, IndexType dim>
+template <typename... Args>
+Array<T, dim>::Array(Args... args)
+  : m_allocator_id(axom::getDefaultAllocatorID())
+  , m_dims {static_cast<IndexType>(args)...}
+{
+  static_assert(sizeof...(Args) == dim,
+                "Array size must match number of dimensions");
+  initialize(detail::packProduct(args...), 0);
+  // Row-major
+  m_strides[dim - 1] = 1;
+  for(int i = static_cast<int>(dim) - 2; i >= 0; i--)
+  {
+    m_strides[i] = m_strides[i + 1] * m_dims[i + 1];
+  }
+}
```

Comment:
Kokkos supported was different layouts (e.g. row-major, column-major, blocking/tiling). This would often facilitate better caching / memory coalescence when executing a common lambda function on different devices.

It's not a necessity, but might be worth considering in the longer-term.

Reply:
Logged in #633